include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -371,113 +346,163 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _
+        auto __kernel_1 =
+            __internal::__kernel_compiler<_LocalScanKernel>::__compile_kernel(::std::forward<_ExecutionPolicy>(__exec));
+        auto __kernel_2 =
+            __internal::__kernel_compiler<_GroupScanKernel>::__compile_kernel(::std::forward<_ExecutionPolicy>(__exec));
+        auto __wgroup_size_kernel_1 =
+            oneapi::dpl::__internal::__kernel_work_group_size(::std::forward<_ExecutionPolicy>(__exec), __kernel_1);
+        auto __wgroup_size_kernel_2 =
+            oneapi::dpl::__internal::__kernel_work_group_size(::std::forward<_ExecutionPolicy>(__exec), __kernel_2);
+        __wgroup_size = ::std::min({__wgroup_size, __wgroup_size_kernel_1, __wgroup_size_kernel_2});
 #endif
 
-    // Practically this is the better value that was found
-    constexpr decltype(__wgroup_size) __iters_per_witem = 16;
-    auto __size_per_wg = __iters_per_witem * __wgroup_size;
-    auto __n_groups = (__n - 1) / __size_per_wg + 1;
-    // Storage for the results of scan for each workgroup
-    sycl::buffer<_Type> __wg_sums(__n_groups);
+        // Practically this is the better value that was found
+        constexpr decltype(__wgroup_size) __iters_per_witem = 16;
+        auto __size_per_wg = __iters_per_witem * __wgroup_size;
+        auto __n_groups = (__n - 1) / __size_per_wg + 1;
+        // Storage for the results of scan for each workgroup
+        sycl::buffer<_Type> __wg_sums(__n_groups);
 
-    _PRINT_INFO_IN_DEBUG_MODE(__exec, __wgroup_size, __mcu);
+        _PRINT_INFO_IN_DEBUG_MODE(__exec, __wgroup_size, __mcu);
 
-    // 1. Local scan on each workgroup
-    auto __submit_event = __exec.queue().submit([&](sycl::handler& __cgh) {
-        oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2); //get an access to data under SYCL buffer
-        auto __wg_sums_acc = __wg_sums.template get_access<access_mode::discard_write>(__cgh);
-        sycl::accessor<_Type, 1, access_mode::discard_read_write, sycl::access::target::local> __local_acc(
-            __wgroup_size, __cgh);
+        // 1. Local scan on each workgroup
+        auto __submit_event = __exec.queue().submit([&](sycl::handler& __cgh) {
+            oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2); //get an access to data under SYCL buffer
+            auto __wg_sums_acc = __wg_sums.template get_access<access_mode::discard_write>(__cgh);
+            sycl::accessor<_Type, 1, access_mode::discard_read_write, sycl::access::target::local> __local_acc(
+                __wgroup_size, __cgh);
 #if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
-        __cgh.use_kernel_bundle(__kernel_1.get_kernel_bundle());
+            __cgh.use_kernel_bundle(__kernel_1.get_kernel_bundle());
 #endif
-        __cgh.parallel_for<_LocalScanKernel>(
+            __cgh.parallel_for<_LocalScanKernel>(
 #if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
-            __kernel_1,
+                __kernel_1,
```

Comment:
As far as I remember, we discussed on one of tech meeting,  according to the spec there is no such API for `parallel_for`, with a "kernel handle" as the first parameter... It needs to calc wg size, right?
What's our plans in the future regarding it?
..And a kind of stupid general question rhetoric (to the spec)  - why a user have to pass a kernel 3 times? -  as a type, as kernel handle and callable object... :-)

Reply:
We already use new `kernel_bundle` API so we should not care about this issue anymore. Anyway, we are developing long term plan how to get reliable work group size and we will see if we need kernel name for that at the end of the day